mlos_core/mlos_core/optimizers/observations.py

```diff
@@ -0,0 +1,361 @@
+                + " if metadata of prior observation is null"
+            )
+
+    def to_list(self) -> List[Observation]:
+        """
+        Converts the observations to a list of observations.
+
+        Returns
+        -------
+        List[Observation]
+            The list of observations.
+        """
+        return [
+            Observation(
+                config=self.config.loc[idx],
+                score=self.score.loc[idx],
+                context=None if self.context is None else self.context.loc[idx],
+                metadata=None if self.metadata is None else self.metadata.loc[idx],
+            )
+            for idx in self.config.index
+        ]
+
+    def __len__(self) -> int:
+        return len(self.config.index)
+
+    def __iter__(self) -> Iterator[Observations]:
+        for idx in self.config.index:
+            yield Observations(
+                config=self.config.loc[[idx]],
+                score=self.score.loc[[idx]],
+                context=None if self.context is None else self.context.loc[[idx]],
+                metadata=None if self.metadata is None else self.metadata.loc[[idx]],
+            )
+
+    def __repr__(self) -> str:
+        return (
+            f"Observation(config={self.config}, score={self.score},"
+            + " context={self.context}, metadata={self.metadata})"
+        )
+
+    def __eq__(self, other: Any) -> bool:
+        if not isinstance(other, Observations):
+            return False
+
+        if not self.config.equals(other.config):
+            return False
+        if not self.score.equals(other.score):
+            return False
+        if not compare_optional_dataframe(self.context, other.context):
+            return False
+        if not compare_optional_dataframe(self.metadata, other.metadata):
+            return False
+
+        return True
+
+    def __ne__(self, other: Any) -> bool:
+        return not self.__eq__(other)
+
+
+class Suggestion:
+    """
+    A single suggestion for a configuration.
+
+    Attributes
+    ----------
+    config : pd.DataFrame
+        The suggested configuration.
+    """
+
+    def __init__(
+        self,
+        *,
+        config: pd.Series,
+        context: Optional[pd.Series] = None,
+        metadata: Optional[pd.Series] = None,
+    ):
+        self.config = config
+        self.context = context
+        self.metadata = metadata
+
+    def complete(self, score: pd.Series) -> Observation:
+        """
+        Completes the suggestion.
+
+        Parameters
+        ----------
+        score : pd.Series
+            The score metrics observed.
+
+        Returns
+        -------
+        Observation
+            The observation of the suggestion.
+        """
+        return Observation(
+            config=self.config,
+            score=score,
+            context=self.context,
+            metadata=self.metadata,
+        )
+
+    def config_to_configspace(self, space: ConfigurationSpace) -> Configuration:
```

Comment:
```suggestion
    def to_configspace_config(self, space: ConfigurationSpace) -> Configuration:
```

?

Since it's actually a Suggestion atm.